src/main/java/net/openhft/chronicle/threads/MediumEventLoop.java

```diff
@@ -192,13 +193,46 @@ public void addHandler(@NotNull final EventHandler handler) {
             }
             throw new IllegalStateException(name() + ": Unexpected priority " + priority + " for " + handler);
         }
+        addHandlerInternal(handler);
+    }
 
-        if (thread == null || thread == Thread.currentThread()) {
+    /**
+     * Add a handler in the appropriate way given the thread adding the handler and the state of the loop
+     */
+    protected void addHandlerInternal(@NotNull EventHandler handler) {
+        if (thread == null) {
+            if (!addHandlerBeforeStart(handler)) {
+                addHandlerAfterStart(handler);
+            }
+        } else if (thread == Thread.currentThread()) {
+            // The event loop thread adding a handler to itself
+            addNewHandler(handler);
+        } else {
+            addHandlerAfterStart(handler);
+        }
+    }
+
+    /**
+     * This is the code used when any thread tries to add an event handler before a loop is started
+     */
+    private boolean addHandlerBeforeStart(@NotNull EventHandler handler) {
+        synchronized (addHandlerMutex) {
+            if (thread != null) {
+                // The loop started since the initial check, fall back to after-start behaviour
+                return false;
+            }
             addNewHandler(handler);
-            return;
         }
+        return true;
+    }
+
+    /**
+     * This is the code executed when a non-event-loop thread wants to add a handler on a started loop
+     */
+    private void addHandlerAfterStart(@NotNull EventHandler handler) {
         do {
             if (isStopped()) {
+                Jvm.startup().on(MediumEventLoop.class, "Tried to add handler to stopped event loop: " + handler);
```

Comment:
suggest maybe say in the log that the attempt to add the handler failed

Reply:
Actually I added that to aid troubleshooting, I might make it debug. Will make the message nicer.